src/components/implementation/fork_tests/voter/voter.c

```diff
@@ -0,0 +1,290 @@
+	if (type == ping) {
+		for (i = 0; i < components[0].nreplicas; i++) {
+			if (components[0].replicas[i].spdid == spdid) return &components[0];
+		}
+		return NULL;
+	}
+	else if (type == pong) {
+		for (i = 0; i < components[1].nreplicas; i++) {
+			if (components[1].replicas[i].spdid == spdid) return &components[1];
+		}
+		return NULL;
+	}
+	else {
+		return NULL;
+	}
+}
+
+struct replica_info *get_replica(spdid_t spdid) {
+	struct nmod_comp *component = NULL;
+	int i;
+	replica_type type = map[spdid];
+	if (type == none) return NULL;
+
+	component = get_component(spdid);
+	if (!component) return NULL;
+
+	for (i = 0; i < component->nreplicas; i++) {
+		if (component->replicas[i].spdid == spdid) return &component->replicas[i];
+	}
+
+	return NULL;
+}
+
+struct channel *get_channel(struct nmod_comp *A, struct nmod_comp *B) {
+	// can we simplify this? Always order to and from in increasing order?
+	int i;
+	if (A == NULL || B == NULL) return NULL;
+
+	for (i = 0; i < n_channels; i++) {
+		if ((channels[i].A == A && channels[i].B == B) ||
+		    (channels[i].A == B && channels[i].B == A)) {
+			return &channels[i];
+		}
+	}
+
+	return NULL;
+}
+
+int block_replica(struct replica_info *replica) {
+	if (!replica->thread_id) BUG();
+	printc("Blocking thread %d\n", replica->thread_id);
+	replica->blocked = 1;
+	return sched_block(cos_spd_id(), 0);
+}
+
+int nwrite(spdid_t spdid, replica_type to, int data) {
+	struct replica_info *replica = get_replica(spdid);
+	struct nmod_comp *component = get_component(spdid);
+	struct nmod_comp *to_comp = (to == ping) ? &components[0] : (to == pong) ? & components[1] : NULL;
+	struct channel *c = get_channel(component, to_comp);
+	int ret;
+	int i, j;
+	if (!replica) BUG();
+	if (!c) BUG();
+
+	if (component == c->A) {
+		if (c->have_A2B_data) {
+			printc("This replica has already sent some data down the channel\n");
+			return -1; // whatever, shouldn't happen if blocked
+		}
+		else {
+			c->have_A2B_data = 1;
+			c->A2B_data = data;
+			ret = block_replica(replica);
+			if (ret < 0) printc("Error\n");
+			/* On wakeup */	
+			return 0;
+		}
+	}
+	else if (component == c->B) {
+		if (c->have_B2A_data) {
+			printc("This replica has already sent some data down the channel\n");
+			return -1; // whatever, shouldn't happen if blocked
+		}
+		else {
+			c->have_B2A_data = 1;
+			c->B2A_data = data;
+			ret = block_replica(replica);
+			if (ret < 0) printc("Error\n");
+			/* On wakeup */	
+			return 0;
+		}
+	}
+	else {
+		BUG();
+		return -1; // shut up compiler
+	}
+}
+
+int nread(spdid_t spdid, replica_type from, int data) {
+	struct replica_info *replica = get_replica(spdid);
+	struct nmod_comp *component = get_component(spdid);
+	struct nmod_comp *from_comp = (from == ping) ? &components[0] : (from == pong) ? & components[1] : NULL;
+	struct channel *c = get_channel(component, from_comp);
+	int ret;
+	if (!replica) BUG();
+	if (!c) BUG();
+
+	if (component == c->A) {
```

Comment:
Well, parts of this code are horribly redundant. I managed to take out one redundant piece of code before pushing for the week. Some remain.